java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java

```diff
@@ -0,0 +1,101 @@
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.CaseGroupTree;
+import org.sonar.plugins.java.api.tree.CaseLabelTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.YieldStatementTree;
+
+@Rule(key = "S6205")
+public class SwitchRedundantKeywordCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final String MESSAGE = "Remove this redundant %s.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CASE_GROUP);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    CaseGroupTree caseGroupTree = (CaseGroupTree) tree;
+    boolean isCaseWithArrow = caseGroupTree.labels().stream().noneMatch(CaseLabelTree::isFallThrough);
+    if (isCaseWithArrow) {
+      caseGroupTree.body().stream()
+        .filter(t -> t.is(Tree.Kind.BLOCK))
+        .map(BlockTree.class::cast)
+        .filter(b -> !b.body().isEmpty())
+        .forEach(this::reportRedundantKeywords);
+    }
+  }
+
+  private void reportRedundantKeywords(BlockTree blockTree) {
+    List<StatementTree> body = blockTree.body();
+    int statementsInBody = body.size();
+    StatementTree lastStatement = body.get(statementsInBody - 1);
+
+    if (statementsInBody == 1) {
+      if (lastStatement.is(Tree.Kind.YIELD_STATEMENT)) {
+        SyntaxToken yieldKeyword = ((YieldStatementTree) lastStatement).yieldKeyword();
+        // Yield can never be implicit in a block, still checking it for defensive programming
+        if (yieldKeyword != null) {
+          reportStatementInBlock(yieldKeyword, blockTree, "block and \"yield\"");
+        }
+      } else {
+        reportIssue(blockTree.openBraceToken(),
+          String.format(MESSAGE, "block"),
+          Collections.singletonList(new JavaFileScannerContext.Location("Redundant close brace", blockTree.closeBraceToken())),
+          null);
+      }
+    } else if (lastStatement.is(Tree.Kind.BREAK_STATEMENT)) {
+      if (statementsInBody == 2) {
+        reportStatementInBlock(lastStatement, blockTree, "block and \"break\"");
+      } else {
+        reportIssue(lastStatement, String.format(MESSAGE, "\"break\""));
+      }
+    }
+  }
+
+  private void reportStatementInBlock(Tree statement, BlockTree blockTree, String redundantParts) {
+    reportIssue(statement,
+      String.format(MESSAGE, redundantParts),
+      Arrays.asList(new JavaFileScannerContext.Location("Redundant opening brace", blockTree.openBraceToken()),
+        new JavaFileScannerContext.Location("Redundant closing brace", blockTree.closeBraceToken())),
+      null);

```

Comment:
Is it correct indentation?

Reply:
I'm afraid it is, at least what is suggested by IntelliJ. In any case, I extracted this code into a separate method to make the whole cleaner.